src/RobustAdaptiveMetropolis.jl

```diff
@@ -0,0 +1,205 @@
+
+julia> # Set the seed so get some consistency.
+       Random.seed!(1234);
+
+julia> # Sample!
+       chain = sample(model, RAM(), 10_000; chain_type=Chains, num_warmup=10_000, progress=false, initial_params=zeros(2));
+
+julia> norm(cov(Array(chain)) - [1.0 0.5; 0.5 1.0]) < 0.2
+true
+```
+
+# References
+[^VIH12]: Vihola (2012) Robust adaptive Metropolis algorithm with coerced acceptance rate, Statistics and computing.
+"""
+Base.@kwdef struct RAM{T,A<:Union{Nothing,AbstractMatrix{T}}} <: AdvancedMH.MHSampler
+    "target acceptance rate"
+    α::T=0.234
+    "negative exponent of the adaptation decay rate"
+    γ::T=0.6
+    "initial lower-triangular Cholesky factor"
+    S::A=nothing
+    "lower bound on eigenvalues of the adapted covariance matrix"
+    eigenvalue_lower_bound::T=0.0
+    "upper bound on eigenvalues of the adapted covariance matrix"
+    eigenvalue_upper_bound::T=Inf
+end
+
+# TODO: Should we record anything like the acceptance rates?
+struct RAMState{T1,L,A,T2,T3}
+    x::T1
+    logprob::L
+    S::A
+    logα::T2
+    η::T3
+    iteration::Int
+    isaccept::Bool
+end
+
+AbstractMCMC.getparams(state::RAMState) = state.x
+AbstractMCMC.setparams!!(state::RAMState, x) = RAMState(x, state.logprob, state.S, state.logα, state.η, state.iteration, state.isaccept)
+
+function step_inner(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::RAM,
+    state::RAMState
+)
+    # This is the initial state.
+    f = model.logdensity
+    d = LogDensityProblems.dimension(f)
+
+    # Sample the proposal.
+    x = state.x
+    U = randn(rng, d)
+    x_new = x + state.S * U
+
+    # Compute the acceptance probability.
+    lp = state.logprob
+    lp_new = LogDensityProblems.logdensity(f, x_new)
+    logα = min(lp_new - lp, zero(lp))  # `min` because we'll use this for updating
+    # TODO: use `randexp` instead.
+    isaccept = log(rand(rng)) < logα
+
+    return x_new, lp_new, U, logα, isaccept
+end
+
+function adapt(sampler::RAM, state::RAMState, logα::Real, U::AbstractVector)
+    # Update `
+    Δα = exp(logα) - sampler.α
+    S = state.S
+    # TODO: Make this configurable by defining a more general path.
+    η = state.iteration^(-sampler.γ)
+    ΔS = η * abs(Δα) * S * U / norm(U)
+    # TODO: Maybe do in-place and then have the user extract it with a callback if they really want it.
+    S_new = if sign(Δα) == 1
+        # One rank update.
+        LinearAlgebra.lowrankupdate(Cholesky(S), ΔS).L
+    else
+        # One rank downdate.
+        LinearAlgebra.lowrankdowndate(Cholesky(S), ΔS).L
+    end
+    return S_new, η
+end
+
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::RAM;
+    initial_params=nothing,
+    kwargs...
+)
+    # This is the initial state.
+    f = model.logdensity
+    d = LogDensityProblems.dimension(f)
+
+    # Initial parameter state.
+    x = initial_params === nothing ? rand(rng, d) : initial_params
```

Comment:
But this is using a RWMH as the main kernel, so IMO we're already assuming unconstrained support for this to be valid

Reply:
> But why prefer rand over randn in that case?

Just did it quickly because I have some vague memory that it's generally preferred to do initialisation in a box near 0 for most of the linking transformations (believe this is the moticvation behind `SampleFromUniform` in DPPL, though I think it technically initialses from a cube centered on 0?).

But it's w/e to me here; we need a better way in general for this, so I'll just change it to `randn` 👍